OKTA-535220 - Add tMFA readme changes

moisesolmedo-okta opened the pull request:
### Problem Analysis (Technical)


### Solution (Technical)


### Affected Components


### Steps to reproduce:

Actual result:

Expected result:

### Tests

mauricesharp-okta commented on README.md:
````diff
@@ -140,6 +140,20 @@ enrollments.forEach { enrollment in
 }
 ```
 
+### Enable Transactional MFA (CIBA) capabilities into existing enrollment
+Enables or disables support for CIBA Transactions for the enrollment. By disabling it, your app won't receive CIBA challenges from the Okta backend.
````
A few things:
- Unless it's a specific technical term "Enable <thing> into existing enrollment" does not make sense in English. It could be "Enable <thing> for an existing enrollment", "Enable existing enrollments to use <thing>", or possibly "Enabling <think> in existing enrollments". "into" is to put something into something else, rather than to add a capability to something else (which AFAIK is what this is about.)
- The title implies that the default setting for CIBA is disabled, is that correct (the suggestion below assumes it is correct)?
- The title implies that this is enabling the Okta backend to use CIBA to enroll the device/app in push... The suggestion below assumes that it's actually about enabling CIBA as one way the app is used to verify the user identity when they attempt to access some other service (e.g. a web app.). It if this isn't the case, let me know and help me understand how CIBA is being used.
- The title and location of this item means that this CIBA is used only to enroll the device. I'm going to assume that i(Suggestion 1 assumes that's the case, suggestion 2 assumes that it's about using the app on the device as a consumer device for user authentication on other services.)
- The heading is hard to understand for someone who doesn't have strong technical knowledge of this area. Is there a way to express it in plainer English (see suggestions below)?
- At the moment, the first sentence on line 144 is almost a restatement of the heading ("or disables" is the only difference.) Is that sentence needed (see suggestion below)?

```suggestion
### Enable using your app for Client Initiated Backchannel Authentication (CIBA)
Enable the Okta backend to send CIBA challenges that verify a user's identity to your app. CIBA challenges are disabled by default. The following code shows enabling challenges for each of your apps enrolled 
```

moisesolmedo-okta replied:
Yes I like that suggestion, easier to understand: only thing I would change:

`The following code shows enabling challenges for each of your apps enrolled`
to
`The following code shows enabling challenges for each of your app's enrollments`

Since this is only possible from one app.

mauricesharp-okta replied:
Sounds fine... I still would like to find a way to talk about an app's enrollments in a simpler way, but that's a challenge for the docs ;-).

FeiChen-okta commented on README.md:
```diff
@@ -140,8 +140,8 @@ enrollments.forEach { enrollment in
 
-### Enable Transactional MFA (CIBA) capabilities into existing enrollment
-Enables or disables support for CIBA Transactions for the enrollment. By disabling it, your app won't receive CIBA challenges from the Okta backend.
+### Enable using your app for Client Initiated Backchannel Authentication (CIBA)
+Enable the Okta backend to send CIBA challenges that verify a user's identity to your app. CIBA challenges are disabled by default. The following code shows enabling challenges for each of your app's enrollments.
```
not sure this is accurate `challenges that verify a user's identity`. The CIBA challenge on the SDK does not have any identity. It is only for authorizing the request. See
https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#rfc.section.2

moisesolmedo-okta replied:
We have the `identity` concept in other parts of the Readme:
-`Enable your app to validate the identity of a user for an Okta authenticator that uses Apple Push Notification service (APNs).`

-`Okta Devices SDK supports identity verification using a custom authenticator in an Okta org. Your app interacts with that custom authenticator in three ways:`

Isn't the sdk verifying the identity of a user?

FeiChen-okta replied:
@mauricesharp-okta @moisesolmedo-okta My confusing here is
```
challenges that verify a user's identity to your app
```
The CIBA request is to authorize a transaction from a consuming device not `to your app`.

moisesolmedo-okta replied:
Makes sense, this would be more accurate:

`Enable the Okta backend to send CIBA challenges to authorize requests from your app. CIBA challenges are disabled by default. The following code shows enabling challenges for each of your app's enrollments.`

FeiChen-okta commented:
@tommywong-okta can you also update the android one to be same or similar to iOS. can do this in your next PR

mauricesharp-okta commented:
I thought CIBA sent challenges to authorize an action on some other device. The wording below says “authorize requests from your app”, but I think it’s supposed to be “

Enable your app to respond to CIBA authorization challenges sent by the Okta backend server. CIBA challenges are disabled by default. The following code shows how to enable challenges for each of the enrolled custom authenticators of your app.


> On Oct 25, 2022, at 5:28 PM, Moises Olmedo ***@***.***> wrote:
> 
> This message originated outside your organization.
> 
> 
> 
> 
> @moisesolmedo-okta commented on this pull request.
> 
> In README.md <https://github.com/okta/okta-devices-swift/pull/72#discussion_r1004980130>:
> 
> > +### Enable using your app for Client Initiated Backchannel Authentication (CIBA)
> +Enable the Okta backend to send CIBA challenges that verify a user's identity to your app. CIBA challenges are disabled by default. The following code shows enabling challenges for each of your app's enrollments.
> Makes sense, this would be more accurate:
> 
> Enable the Okta backend to send CIBA challenges to authorize requests from your app. CIBA challenges are disabled by default. The following code shows enabling challenges for each of your app's enrollments.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/okta/okta-devices-swift/pull/72#discussion_r1004980130>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AWKDU3W5KTUIBIF5HY4DKHDWFBGHHANCNFSM6AAAAAARLQ2ITI>.
> You are receiving this because you were mentioned.
>

mauricesharp-okta commented:
So far this talks about enabling CIBA. What about responding to the challenge? Is there anything different a developer needs to do or is it just another challenge that’s handled by this kind of code:

func initOktaDeviceAuthenticator() {
    do {
        let applicationConfig = ApplicationConfig(applicationName: "Your-app-name",
                                                  applicationVersion: "Your-app-version",
                                                  applicationGroupId: "com.your.group.id",
                                                  )
        applicationConfig.approveActionTitle = "Approve"
        applicationConfig.denyActionTitle =  "Deny"
        applicationConfig.userVerificationActionTitle = "Verify in YourAppName"
        ...


> On Oct 25, 2022, at 5:55 PM, Fei Chen ***@***.***> wrote:
> 
> This message originated outside your organization.
> 
> 
> 
> 
> @tommywong-okta <https://github.com/tommywong-okta> can you also update the android one to be same or similar to iOS. can do this in your next PR
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/okta/okta-devices-swift/pull/72#issuecomment-1291185378>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AWKDU3U6T22J4HMFL474MDTWFBJOLANCNFSM6AAAAAARLQ2ITI>.
> You are receiving this because you were mentioned.
>

moisesolmedo-okta commented:
@mauricesharp-okta There's no change for developers to respond to the challenge, same as before. 
Only difference is we added a new challenge protocol in case they want to retrieve the CIBA `binding message`. 
Added an example on SampleApp, following the same pattern we've been using to get other properties from the challenge (such as client location, OS, date, etc):
https://github.com/okta/okta-devices-swift/blob/master/Examples/PushSampleApp/SampleApp/UserConsent/UserConsentViewModel.swift#L79

mauricesharp-okta commented:
Thanks @moisesolmedo-okta. @tommywong-okta , please include me on the PR for the Android wording. Once I have that I can update the existing custom authenticator docs with this option.
